Adds support for bilinear interpolation of frames

Philbrick opened the pull request:
Implements modified bilinear interpolation as an optional method for  improved downsampling.

Philbrick committed:
Adds support for bilinear interpolation of frames

Implements modified bilinear interpolation as an optional method for  improved downsampling.

danielbeaudreau commented:
/gcbrun

danielbeaudreau commented on src/bilinearinterpolationframe.cpp:
```diff
@@ -0,0 +1,288 @@
+
+  if (openslide_get_error(osr_)) {
+    BOOST_LOG_TRIVIAL(error) << openslide_get_error(osr_);
+  }
```
should we throw an exception here?

Philbrick replied:
sure added a throw

danielbeaudreau commented on src/bilinearinterpolationframe.cpp:
```diff
@@ -0,0 +1,288 @@
+  const double TargetFrame_YStart =
+      static_cast<double>(targetLevelHeight_ * locationY_ / levelHeight_);
+
+  // Cache X coordinate position trannsformation to transform from layer being
+  // downsampled to frame coordinates. Performance
```
incomplete comment

danielbeaudreau commented on src/bilinearinterpolationframe.h:
```diff
@@ -0,0 +1,78 @@
+#include "src/enums.h"
+#include "src/frame.h"
+
+// Frame represents a single image frame from the OpenSlide library
+class BilinearInterpolationFrame : public Frame {
+ public:
+  // osr - openslide
+  // locationX, locationY - top-left corner of frame in level coordinates
+  // frameWidhtDownsampled, frameHeightDownsampled - size of frame to get
+  // multiplicator - size difference between 0 level and current one
+  // frameWidht, frameHeight_ - size frame needs to be scaled to
+  // compression - type of compression
```
Comments are out of date and contain typos

danielbeaudreau commented on src/bilinearinterpolationframe.cpp:
```diff
@@ -0,0 +1,288 @@
+
+void set_pixel(double *rgb_mem, const int width, const int height, int cx,
+               int cy, const double red, const double green, const double blue,
+               const double precent) {
```
typo: percent

Philbrick replied:
Fixed

danielbeaudreau commented on tests/bilinearinterpolationframe.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "src/bilinearinterpolationframe.h"
+#include "tests/testUtils.h"
+
+TEST(BilinearInterpolationFrame, jpeg) {
+  openslide_t* osr = openslide_open(tiffFileName);
+  BilinearInterpolationFrame frame(osr, 0, 0, 0, 100, 100, 100, 100, JPEG, 1,
+                                   500, 500, 1000, 1000, 2000, 2000);
+  frame.sliceFrame();
+  ASSERT_TRUE(frame.isDone());
+  ASSERT_GE(frame.getSize(), 0);
+}
+
+TEST(BilinearInterpolationFrame, jpeg2000Scaling) {
+  openslide_t* osr = openslide_open(tiffFileName);
+  BilinearInterpolationFrame frame(osr, 0, 0, 0, 1000, 1000, 100, 100, JPEG2000,
+                                   1, 500, 500, 1000, 1000, 2000, 2000);
+  frame.sliceFrame();
+  ASSERT_TRUE(frame.isDone());
+  ASSERT_GE(frame.getSize(), 0);
```
Could we add some end to end test with golden file diffs so that we can ensure that any subsequent changes to the bilinear downsampling code does not break anything?

Philbrick committed:
Fix downsampeling layer selection and dimensions

* Fixes downsampling layer selection, selecting slide which is closest match for target downsample resolution
* Correction for Aperio reported non integer downsamples.
* Adjusts dimensions of downsampled image to account for dimensions lost due to downsampleing.

danielbeaudreau commented:
could we keep the downsampling dimension fix in a separate PR to make this easier to review?

Philbrick committed:
additional test cases

danielbeaudreau commented:
looks good. please revert the dimensions changes here and submit a separate pr, then I will merge this

Philbrick committed:
Rolling back changes

Rolling back to reduce size of c "Fix downsampeling layer selection and dimensions" CL.  CL is fine and will be resubmitted.

danielbeaudreau commented:
/gcbrun

danielbeaudreau commented:
could you raise the timeout in wsi-to-dicom-converter/endToEnd/endToEndTest.yaml to 3600s so the presubmit does not time out?

Philbrick committed:
Increased timeout

danielbeaudreau commented:
/gcbrun

danielbeaudreau commented:
/gcbrun

danielbeaudreau commented:
/gcbrun
